presto-benchmark/src/main/java/com/facebook/presto/benchmark/SqlRewriteConditionalAggregationBenchmarks.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.benchmark;
+
+import com.facebook.presto.testing.LocalQueryRunner;
+import com.google.common.collect.ImmutableMap;
+
+import static com.facebook.presto.benchmark.BenchmarkQueryRunner.createLocalQueryRunner;
+
+public class SqlRewriteConditionalAggregationBenchmarks
+        extends AbstractSqlBenchmark
+{
+    public SqlRewriteConditionalAggregationBenchmarks(LocalQueryRunner localQueryRunner, String sql)
+    {
+        super(localQueryRunner, "sql_rewrite_conditional_aggregation", 5, 10, sql);
+    }
+
+    public static void main(String[] args)
```

Comment:
I have run the benchmark. But there is no major difference in performance. Can we check why?

```
Without optimization
Oct 07, 2022 10:07:55 PM com.facebook.airlift.log.Logger info
INFO: -- Loading system access control --
Oct 07, 2022 10:07:55 PM com.facebook.airlift.log.Logger info
INFO: -- Loaded system access control allow-all --
Oct 07, 2022 10:07:55 PM com.facebook.airlift.log.Logger info
INFO: -- Loading temp storage local --
Oct 07, 2022 10:07:55 PM com.facebook.airlift.log.Logger info
INFO: -- Loaded temp storage local --
Oct 07, 2022 10:07:55 PM com.facebook.airlift.log.Logger info
INFO: -- Loading temp storage local --
Oct 07, 2022 10:07:55 PM com.facebook.airlift.log.Logger info
INFO: -- Loaded temp storage local --
peak_memory:39834288,elapsed_millis:984,input_rows_per_second:61177,output_rows_per_second:207813,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:983613209,cpu_nanos:913704000,user_nanos:856376000,input_rows:60175,input_bytes:0,output_rows:204408,output_bytes:7325046
peak_memory:39834288,elapsed_millis:970,input_rows_per_second:62061,output_rows_per_second:210817,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:969596041,cpu_nanos:910932000,user_nanos:861031000,input_rows:60175,input_bytes:0,output_rows:204408,output_bytes:7325046
peak_memory:39834288,elapsed_millis:1169,input_rows_per_second:51453,output_rows_per_second:174783,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:1169495125,cpu_nanos:1080444000,user_nanos:972766000,input_rows:60175,input_bytes:0,output_rows:204408,output_bytes:7325046
peak_memory:39834288,elapsed_millis:1018,input_rows_per_second:59082,output_rows_per_second:200697,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:1018490542,cpu_nanos:967825000,user_nanos:917881000,input_rows:60175,input_bytes:0,output_rows:204408,output_bytes:7325046
peak_memory:39834288,elapsed_millis:929,input_rows_per_second:64793,output_rows_per_second:220096,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:928718416,cpu_nanos:874630000,user_nanos:826043000,input_rows:60175,input_bytes:0,output_rows:204408,output_bytes:7325046
peak_memory:39834288,elapsed_millis:1029,input_rows_per_second:58491,output_rows_per_second:198689,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:1028781541,cpu_nanos:982264000,user_nanos:929156000,input_rows:60175,input_bytes:0,output_rows:204408,output_bytes:7325046
peak_memory:39834288,elapsed_millis:1118,input_rows_per_second:53838,output_rows_per_second:182882,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:1117701833,cpu_nanos:1047757000,user_nanos:937719000,input_rows:60175,input_bytes:0,output_rows:204408,output_bytes:7325046
peak_memory:39834288,elapsed_millis:849,input_rows_per_second:70862,output_rows_per_second:240713,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:849176458,cpu_nanos:815435000,user_nanos:783957000,input_rows:60175,input_bytes:0,output_rows:204408,output_bytes:7325046
peak_memory:39834288,elapsed_millis:956,input_rows_per_second:62952,output_rows_per_second:213841,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:955886459,cpu_nanos:906190000,user_nanos:858392000,input_rows:60175,input_bytes:0,output_rows:204408,output_bytes:7325046
peak_memory:39834288,elapsed_millis:1027,input_rows_per_second:58600,output_rows_per_second:199060,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:1026864250,cpu_nanos:960450000,user_nanos:892033000,input_rows:60175,input_bytes:0,output_rows:204408,output_bytes:7325046
sql_rewrite_conditional_aggregation ::  945.963 cpu ms ::  38MB peak memory :: in 60.2K,      0B,   63.6K/s,      0B/s :: out  204K,  6.99MB,    216K/s,  7.38MB/s



With optimization
Oct 07, 2022 10:08:22 PM com.facebook.airlift.log.Logger info
INFO: -- Loading system access control --
Oct 07, 2022 10:08:22 PM com.facebook.airlift.log.Logger info
INFO: -- Loaded system access control allow-all --
Oct 07, 2022 10:08:22 PM com.facebook.airlift.log.Logger info
INFO: -- Loading temp storage local --
Oct 07, 2022 10:08:22 PM com.facebook.airlift.log.Logger info
INFO: -- Loaded temp storage local --
Oct 07, 2022 10:08:22 PM com.facebook.airlift.log.Logger info
INFO: -- Loading temp storage local --
Oct 07, 2022 10:08:22 PM com.facebook.airlift.log.Logger info
INFO: -- Loaded temp storage local --
peak_memory:39834288,elapsed_millis:982,input_rows_per_second:61295,output_rows_per_second:208214,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:981719625,cpu_nanos:884693000,user_nanos:817050000,input_rows:60175,input_bytes:0,output_rows:204408,output_bytes:7325046
peak_memory:39834288,elapsed_millis:825,input_rows_per_second:72948,output_rows_per_second:247796,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:824901792,cpu_nanos:790391000,user_nanos:758429000,input_rows:60175,input_bytes:0,output_rows:204408,output_bytes:7325046
peak_memory:39834288,elapsed_millis:959,input_rows_per_second:62725,output_rows_per_second:213072,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:959333500,cpu_nanos:905687000,user_nanos:854889000,input_rows:60175,input_bytes:0,output_rows:204408,output_bytes:7325046
peak_memory:39834288,elapsed_millis:980,input_rows_per_second:61406,output_rows_per_second:208589,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:979953000,cpu_nanos:934085000,user_nanos:884335000,input_rows:60175,input_bytes:0,output_rows:204408,output_bytes:7325046
peak_memory:39834288,elapsed_millis:1035,input_rows_per_second:58130,output_rows_per_second:197461,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:1035176667,cpu_nanos:960926000,user_nanos:865272000,input_rows:60175,input_bytes:0,output_rows:204408,output_bytes:7325046
peak_memory:39834288,elapsed_millis:806,input_rows_per_second:74649,output_rows_per_second:253575,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:806102250,cpu_nanos:771425000,user_nanos:741661000,input_rows:60175,input_bytes:0,output_rows:204408,output_bytes:7325046
peak_memory:39834288,elapsed_millis:820,input_rows_per_second:73410,output_rows_per_second:249368,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:819702125,cpu_nanos:780943000,user_nanos:748123000,input_rows:60175,input_bytes:0,output_rows:204408,output_bytes:7325046
peak_memory:39834288,elapsed_millis:958,input_rows_per_second:62785,output_rows_per_second:213274,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:958427916,cpu_nanos:917175000,user_nanos:875293000,input_rows:60175,input_bytes:0,output_rows:204408,output_bytes:7325046
peak_memory:39834288,elapsed_millis:1032,input_rows_per_second:58328,output_rows_per_second:198134,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:1031664708,cpu_nanos:953599000,user_nanos:892026000,input_rows:60175,input_bytes:0,output_rows:204408,output_bytes:7325046
peak_memory:39834288,elapsed_millis:987,input_rows_per_second:60977,output_rows_per_second:207135,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:986831917,cpu_nanos:934477000,user_nanos:886629000,input_rows:60175,input_bytes:0,output_rows:204408,output_bytes:7325046
sql_rewrite_conditional_aggregation ::  883.340 cpu ms ::  38MB peak memory :: in 60.2K,      0B,   68.1K/s,      0B/s :: out  204K,  6.99MB,    231K/s,  7.91MB/s

Process finished with exit code 0

Reply:
I think it's because that the aggregation only consists a small portion of the total cost, and the saving from aggregating on fewer columns is not significant enough for this query.

Just updated the PR and added a new test query to verify that we indeed do less aggregation computation in AbstractTestQueries.java.
```
    @Test
    public void testRewriteConditionalAggregation()
    {
        String sql = "select y, grouping(y) grouping_y, IF(grouping(y)!=1, sum(x)) from (VALUES (1,1,1), (9223372036854775807,2,1)) AS T(x,y,z)"
                + " group by grouping sets ((y), (z)) ORDER BY 1, 2";
        assertQueryFails(sql, "bigint addition overflow:.*");
        Session rewriteConditionalAggregation = Session.builder(getSession())
                .setSystemProperty(REWRITE_CONDITIONAL_AGGREGATION_ENABLED, "true")
                .build();
        assertQuery(rewriteConditionalAggregation, sql, getSession(), "VALUES (1, 0, 1), (2, 0, 9223372036854775807), (NULL, 1, NULL)");
    }
```

Reply:
Updated the benchmark to include more test queries.

The improvement from this optimization depends on two aspects: 1) how many aggregation computations we skip by pushing the condition to aggregation, 2) how expensive the aggregation is.

To demonstrate how effective this optimization can be, we first add a query which has a cube operator, and use the expensive approx_distinct function. For this query, the CPU time decrease from 665ms to 432 ms.

Then I remove the cube operator but only keep 3 grouping sets (and still use the approx_distinct aggregation), the CPU time decrease from 232 ms to 206 ms.

Then I change the aggregation from approx_distinct to sum which is cheaper (and keeping the cube), the CPU time decreases from 443ms to 414 ms.

Considering that there are to be many grouping sets and expensive UDAF used according to our customer, I expect this optimization to show more savings. cc @highker @kaikalur